src/Core/Grid/Data/Factory/AttachmentGridDataFactoryDecorator.php

```diff
@@ -0,0 +1,160 @@
+        int $employeeIdLang,
+        Connection $connection,
+        string $dbPrefix,
+        FileSizeConverter $fileSizeConverter
+    ) {
+        $this->attachmentDoctrineGridDataFactory = $attachmentDoctrineGridDataFactory;
+        $this->employeeIdLang = $employeeIdLang;
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
+        $this->fileSizeConverter = $fileSizeConverter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData(SearchCriteriaInterface $searchCriteria)
+    {
+        $attachmentData = $this->attachmentDoctrineGridDataFactory->getData($searchCriteria);
+
+        $attachmentRecords = $this->applyModifications($attachmentData->getRecords());
+
+        return new GridData(
+            $attachmentRecords,
+            $attachmentData->getRecordsTotal(),
+            $attachmentData->getQuery()
+        );
+    }
+
+    /**
+     * @param RecordCollectionInterface $attachments
+     *
+     * @return RecordCollection
+     */
+    private function applyModifications(RecordCollectionInterface $attachments): RecordCollection
+    {
+        $modifiedAttachments = [];
+
+        foreach ($attachments as $attachment) {
+            if ((int) $attachment['products'] > 0) {
+                $productNamesArray = $this->getProductNames($attachment['id_attachment']);
+                $productNames = $productNamesArray['product_names'] ?? '';
+                $attachment['dynamic_message'] = $this->trans(
+                    'This file is associated with the following products. Are you sure you want to delete it?',
+                    [],
+                    'Admin.Catalog.Notification'
+                );
+                $attachment['dynamic_message'] .= PHP_EOL . PHP_EOL . $productNames;
+            }
+
+            $attachment['file_size'] = $this->fileSizeConverter->convert((int) $attachment['file_size']);
+            $attachment['products'] .= ' ' . $this->trans('product(s)', [], 'Admin.Catalog.Feature');
+
+            $modifiedAttachments[] = $attachment;
+        }
+
+        return new RecordCollection($modifiedAttachments);
+    }
+
+    /**
+     * @param string $attachmentId
+     *
+     * @return array
+     */
+    private function getProductNames(string $attachmentId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+
+        $qb->select("GROUP_CONCAT(DISTINCT pl.`name` SEPARATOR ', ') as product_names")
+            ->from($this->dbPrefix . 'product_attachment', 'pa')
+            ->leftJoin(
+                'pa',
+                $this->dbPrefix . 'product_lang',
+                'pl',
+                'pa.`id_product` = pl.`id_product` AND pl.`id_lang` = :langId'
+            )
+            ->where('pa.`id_attachment` = :attachmentId')
+            ->setParameter('attachmentId', $attachmentId)
+            ->setParameter('langId', $this->employeeIdLang)
+            ->setMaxResults(1);
+
+        return $qb->execute()->fetch();
```

Comment:
```suggestion
        return $qb->getSingleResult();
```

ANd you can remove the setMaxResults.

But the function name is awkward, you return a string containing all names.
Why not returning an array[] of names instead :thinking: ?

Reply:
There is no getSingleResult method in QueryBuilder but i agree about returning array so $qb->execute()->fetchAll(PDO::FETCH_COLUMN); will suffice i guess :)